internal/validation/dataplane/validation.go

```diff
@@ -0,0 +1,194 @@
+	}
+	return nil
+}
+
+// getDBModeFromEnv gets the dbmode from Env.
+// If the second return value is false, the dbMode is not found in Env.
+func (v *Validator) getDBModeFromEnv(namespace string, envs []corev1.EnvVar) (string, bool, error) {
+
+	dbMode := ""
+	dbModeFound := false
+	for _, envVar := range envs {
+		// use the last appearance of the same key as the result since k8s takes this precedence.
+		if envVar.Name == consts.EnvVarKongDatabase {
+			// value is non-empty.
+			if envVar.Value != "" {
+				dbMode = envVar.Value
+				dbModeFound = true
+			} else if envVar.ValueFrom != nil {
+				// value is empty,get from ValueFrom from configmap/secret.
+				if envVar.ValueFrom.ConfigMapKeyRef != nil {
+					var err error
+					dbMode, dbModeFound, err = v.getValueFromConfigMapKeyRef(namespace, envVar.ValueFrom.ConfigMapKeyRef)
+					if err != nil {
+						return "", false, err
+					}
+				}
+				if envVar.ValueFrom.SecretKeyRef != nil {
+					var err error
+					dbMode, dbModeFound, err = v.getValueFromSecretRef(namespace, envVar.ValueFrom.SecretKeyRef)
+					if err != nil {
+						return "", false, err
+					}
+				}
+			}
+		}
+	}
+	return dbMode, dbModeFound, nil
+}
+
+func (v *Validator) getDBModeFromEnvFrom(namespace string, envFroms []corev1.EnvFromSource) (string, bool, error) {
+	dbMode := ""
+	dbModeFound := false
+	for _, envFrom := range envFroms {
+		// if the envFrom.Prefix is the prefix of KONG_DATABASE,
+		// it is possible that this envFrom contains values of KONG_DATABASE.
+		if strings.HasPrefix(consts.EnvVarKongDatabase, envFrom.Prefix) {
+			if envFrom.ConfigMapRef != nil {
+				var err error
+				dbMode, dbModeFound, err = v.getDBModeFromConfigMapRef(namespace, envFrom.Prefix, envFrom.ConfigMapRef)
+				// technically it goes slightly against eventual-consistency to throw an error here,
+				// but the alternative is that we would need to validate ALL ConfigMaps on create
+				// and do relational mapping to DataPlane resources to validate that they aren't
+				// going to introduce a new violation, or we would have to do an additional level
+				// of validation that could only run during reconciliation.
+				if err != nil {
+					return "", false, err
+				}
+			}
+			if envFrom.SecretRef != nil {
+				var err error
+				dbMode, dbModeFound, err = v.getDBModeFromSecretRef(namespace, envFrom.Prefix, envFrom.SecretRef)
+				if err != nil {
+					return "", false, err
+				}
+			}
+		}
+	}
+	return dbMode, dbModeFound, nil
+}
+
+func (v *Validator) getValueFromConfigMapKeyRef(namespace string, cmKeyRef *corev1.ConfigMapKeySelector) (string, bool, error) {
+	cm := &corev1.ConfigMap{}
+	namespacedName := k8stypes.NamespacedName{Namespace: namespace, Name: cmKeyRef.Name}
+	err := v.c.Get(context.Background(), namespacedName, cm)
+	if err != nil {
+		return "", false, fmt.Errorf("failed to get configMap %s in configMapKeyRef, error %w", cmKeyRef.Name, err)
```

Comment:
commonly we use a colon to separate the messaging about the error from the error itself in a wrapped error as this usually ends up being a pretty clear delineation:

```suggestion
		return "", false, fmt.Errorf("failed to get configMap %s in configMapKeyRef: %w", cmKeyRef.Name, err)
```

Reply:
updated, and the same for the following.